File: examples/simple/src/main/scala/com/example/hello/MainApp.scala
package com.example.hello

import scala.ioc.ppm._
import scala.ioc.Factory

import java.time.ZonedDateTime

object MainApp extends App {

  // build the factory and staff it
  val (factory, _) = staffFactoryFromResource("staff.fsp")
  // put the "helloWorld" manager to work and get the fruits of her labor
  val obj = factory.putToWork("helloWorld").asInstanceOf[HelloWorld]
  obj.printMessage

  // gimme the time, ya lazy bum!
  val dateTime = factory.putToWork("lazyManager").asInstanceOf[ZonedDateTime]
  // you, people-pleaser, gimme the time, too.
  val anotherDateTime = factory.putToWork("eagerManager").asInstanceOf[ZonedDateTime]

  // let's wait half a second
  Thread sleep 500

  // who will give the same ol', and who will give something new?
  println(s"Will the lazy manager produce the same as before?"
      + s" ${if (factory.putToWork("lazyManager").equals(dateTime)) "Yes" else "No"}")
  println(s"Will the eager manager produce the same as before?"
      + s" ${if (factory.putToWork("eagerManager").equals(anotherDateTime)) "Yes" else "No"}")

  // The productive senior manager puts the lazy one to work.
  val dateTimeHelloWorld =
    factory.putToWork("productiveSeniorManager").asInstanceOf[DateTimeHelloWorld]
  dateTimeHelloWorld.printMessage
  // Will the lazy manager give the productive one the same ol'?
  println("The senior manager will make even the lazy manager perform some work.")
  println(s"Will the lazy manager give senior manager the same as before?"
      + s" ${if (dateTimeHelloWorld.dateTime.equals(dateTime)) "Yes" else "No"}")
}

File: examples/java/src/main/java/com/example/hello/MainApp.java
package com.example.hello;

import java.time.ZonedDateTime;
import java.util.function.Supplier;

public class MainApp {

  public static void run(
      Supplier<HelloWorld> obj0,
      Supplier<ZonedDateTime> dateTime0,
      Supplier<ZonedDateTime> anotherDateTime0,
      Supplier<DateTimeHelloWorld> dateTimeHelloWorld0
  ) {

    obj0.get().printMessage();

    // gimme the time, ya lazy bum!
    final ZonedDateTime dateTime = dateTime0.get();
    // you, people-pleaser, gimme the time, too.
    final ZonedDateTime anotherDateTime = anotherDateTime0.get();

    // let's wait half a second
    try {
        Thread.sleep(500);
    }
    catch(InterruptedException e) {
        throw new RuntimeException(e);
    }

    // who will give the same ol', and who will give something new?
    System.out.println("Will the lazy manager produce the same as before?"
        + " " + (dateTime0.get().equals(dateTime) ? "Yes" : "No"));
    System.out.println("Will the eager manager produce the same as before?"
        + " " + (anotherDateTime0.get().equals(anotherDateTime) ? "Yes" : "No"));

    // The productive senior manager puts the lazy one to work.
    final DateTimeHelloWorld dateTimeHelloWorld = dateTimeHelloWorld0.get();
    dateTimeHelloWorld.printMessage();
    // Will the lazy manager give the productive one the same ol'?
    System.out.println("The senior manager will make even the lazy manager perform some work.");
    System.out.println("Will the lazy manager give senior manager the same as before?"
        + " " + (dateTimeHelloWorld.dateTime.equals(dateTime) ? "Yes" : "No"));
  }
}
